create volunteer-headshot component

WenInCode opened the pull request:
# What's in this PR?
## Changes
- Created `volunteer-headshot` component
- Created `volunteer-headshot` page-object
- Added tests 

## Todo
- add styling
- use the component in the thankyou page.

## References
Fixes #678

JoshSmith commented:
@WenInCode think maybe there's a misunderstanding here. We want to take one of the roles out of the user's own `user-roles` and use that for display.

WenInCode commented:
@JoshSmith ha that makes so much more sense

JoshSmith commented:
@WenInCode sorry! Did I not explain it well enough in the issue?

JoshSmith commented:
I think I didn't. Was pretty vague.

JoshSmith commented:
@WenInCode is this reviewable now?

WenInCode commented:
@JoshSmith negative, working on putting it on the thank-you page and need to add styling. Working on it right now, but I will likely finish tomorrow as it's pretty late.

WenInCode commented:
I didn't get to work as much as I had hoped to over the past two days, but here is some progress (I only had 1 contributor so I hard coded more). There are still some things that need to be finished.

- styles
- selecting subset of contributors

It has been fun to try and do some scss even if I am not great at it yet. Sorry this is taking longer than expected.

<img width="1440" alt="code_corps" src="https://cloud.githubusercontent.com/assets/5448834/20782551/f9b2f5b0-b749-11e6-95bc-f9c7da905d61.png">

JoshSmith commented:
@WenInCode this is _great_ progress. Also, you look like a pretty prolific person there! Backend Developer and Tester and Data Scientist. 🔥 🔥 🔥 🚒

WenInCode commented:
Updated a bunch of things. The volunteer list loads a bit slowly because the `userRoles` or `roles` (forgot which) aren't side loaded. 

One thing I have not figured out is how to center-align the list of contributors. I could use some help there. Otherwise, I think this is ready for review.

Tests are passing locally, not sure why they were killed in circle.

marineb commented:
@WenInCode I'd be happy to help center align the contributors (though I'm not sure how to checkout your branch). Let me know if you still want someone to chime in.

sbatson5 commented on app/components/volunteer-headshot.js:
```diff
@@ -0,0 +1,98 @@
+    @property volunteerName
+    @type String
+   */
+  volunteerName: computed('volunteer.name', 'volunteer.firstName', 'volunteer.lastName', function() {
```
If we are observing multiple properties on 1 object, we can simplify what we are observing to:
`computed('volunteer.{name,firstName,lastName}', ...`

sbatson5 commented on app/templates/components/volunteer-headshot.hbs:
```diff
@@ -0,0 +1,9 @@
+    <p class="volunteer-headshot__role" data-test-selector="volunteer role">{{userRole.role.name}}</p>
+  {{else}}
+    <p class="volunteer-headshot__role" data-test-selector="volunteer role">Volunteer</p>
+  {{/if}}
```
Rather than doing the `if/else` block, it looks like we can get away with an inline `if` since we are only updating 1 item.

```
<p class="volunteer-headshot__role" data-test-selector="volunteer role">
  {{if userRole userRole.role.name "Volunteer"}}
</p>
```

WenInCode replied:
I completely forgot about the inline if 😜 - thanks!

sbatson5 commented on tests/integration/components/volunteer-headshot-test.js:
```diff
@@ -0,0 +1,73 @@
+  'Designer & Developer'
+];
+
+function generateUserRoles() {
```
Since we are just adding 1 object per role to this array (i.e. the number of roles never varies), thoughts on just making it static to start with:

```
const usersRoles = [
  { role: { name: 'Developer' } },
  { role: { name: 'Ember Developer' } },
  { role: { name: 'UX Designer' } },
  { role: { name: 'Software Engineer' } },
  { role: { name: 'Project Coordinator' } },
  { role: { name: 'Designer & Developer } }'
];
```
This way we aren't rebuilding this on every test.

WenInCode commented on app/components/thank-you-container.js:
```diff
@@ -1,9 +1,101 @@
+        return get(orgMembership, 'member');
+      });
+
+      if (volunteers.length > 12) {
```
Ya that sounds good to me. Normally when I'm working with magic numbers I would remember to do this! Good catch 👍

WenInCode commented:
Thanks @sbatson5 for the feedback!

@marineb help would be greatly appreciated! The way I work with someone else's fork is the following:

## add the remote 
```
git remote add [whatever you want to call the remote] [the remotes url]

# ex.
git remote add wenincode https://github.com/wenincode/code-corps-ember
```

## fetch the changes
```
git fetch wenincode 

# or get all
git fetch --all
```

## checkout branches
and then you chould be able to checkout into the branch.

```
git checkout add-volunteer-headshot-component
```

## push changes
When your ready just push to your changes to my remote (I will give you commit access to my repo)
```
git push wenincode add-volunteer-headshot-component
```

*I have never actually pushed changes to someones remote, but this is the flow I normally use if I want to pull down someones changes.*

WenInCode commented:
@sbatson5 I've addressed the feedback you've given - thanks again! 
Putting this back into review.

marineb commented:
@WenInCode I wasn't able to push directly to your codebase, so I sent you a [PR](https://github.com/WenInCode/code-corps-ember/pull/1).

JoshSmith commented on app/components/thank-you-container.js:
```diff
@@ -1,9 +1,103 @@
+    @property nonPendingMemberships
+    @type Ember.Array
+   */
+  nonPendingMemberships: computed('organizationMemberships.[]', 'organizationMemberships.@each.role', function() {
```
What is the reason for doing `.[]` and `.@each.role`?

WenInCode replied:
@JoshSmith I had an idea of why I was doing this, but I wasn't sure I was right. So I threw together this [twiddle to show the differences](https://ember-twiddle.com/5596716e1499fb3cb86c77806222800e?openFiles=templates.components.computed-each-attr.hbs%2Ctemplates.components.computed-each-attr.hbs).

- computed property on an array of objects won't recompute on things being pushed/popped from the array if `obj.[]` is not used.
- computed property on an array of objects won't recompute if an object in that arrays attributes change without `@each.thatattribute`

However, I agree with your statement below, so this will probably be taken out. But I figured I should still investigate.

JoshSmith replied:
@WenInCode I figured this was the case and I reviewed top to bottom here, so the other comments are laggard comments.

JoshSmith commented on app/components/thank-you-container.js:
```diff
@@ -1,9 +1,103 @@
+
+    if (isPresent(memberships)) {
+      return memberships.filter((membership) => {
+        let isPending = get(membership, 'isPending');
```
The `organization` model has a line `pendingMemberships: filterBy('organizationMemberships', 'isPending')`. Maybe we could have an `approvedMemberships` and then you can just compute on the `approvedMemberships`?

WenInCode replied:
agreed. one thing is that the role has to be fetched, so there is a portion of time where `role` on the memberships is `undefined`. During that loading, all memberships with `undefined` `role`s will be not `pending`.

I guess what I am getting at is the cleanest way to do it is (and the way I have changed it to):
`approvedMemberships: computed.filterBy('organizationMemberships', 'isNotPending'),`
but that does not filter properly when `role` is undefined.

JoshSmith commented on app/components/thank-you-container.js:
```diff
@@ -1,9 +1,103 @@
+    if (isPresent(nonPendingMemberships)) {
+      let volunteers = nonPendingMemberships.map((orgMembership) => {
+        return get(orgMembership, 'member');
+      });
```
You can do `Ember.computed.mapBy` instead. See above for how to put an `approvedMembers` directly on the `organization` model.

JoshSmith commented on app/components/thank-you-container.js:
```diff
@@ -1,9 +1,103 @@
+  volunteers: computed('nonPendingMemberships', function() {
+    let nonPendingMemberships = get(this, 'nonPendingMemberships');
+
+    if (isPresent(nonPendingMemberships)) {
```
Do you need `isPresent` here?

JoshSmith commented on app/components/volunteer-headshot.js:
```diff
@@ -0,0 +1,98 @@
+      'volunteer.userRoles.@each.user',
+    function() {
+      return this.get('volunteer.userRoles');
+    }),
```
Why not just an `alias` here?

JoshSmith commented on app/components/volunteer-headshot.js:
```diff
@@ -0,0 +1,98 @@
+    } else if (isPresent(firstName) && isPresent(lastName)) {
+      return `${firstName} ${lastName}`;
+    } else {
+      return '';
```
In the `else` would fallback to their username, which we know must be set.

JoshSmith commented on app/components/volunteer-headshot.js:
```diff
@@ -0,0 +1,98 @@
+    @property isVolunteerNamePresent
+    @type Boolean
+   */
+  isVolunteerNamePresent: computed.notEmpty('volunteerName'),
```
Can remove this if you implement the above comment.

WenInCode replied:
ya I had originally put it because I was getting some fun `undefined undefined` as names, because the members promise hadn't resolved.

JoshSmith commented on app/components/volunteer-headshot.js:
```diff
@@ -0,0 +1,98 @@
+  altText: computed('volunteerName', function() {
+    let name = get(this, 'volunteerName');
+
+    return `${name}\'s headshot`;
```
Would remove "headshot" here, and let the alt text just be their name.

JoshSmith commented on tests/integration/components/volunteer-headshot-test.js:
```diff
@@ -0,0 +1,63 @@
+  beforeEach() {
+    this.set('user', {
+      name: 'Test User',
+      photoThumbUrl: 'http://fillmurray.com/200/200',
```
Can we use a URL here that doesn't make an extra HTTP request?

WenInCode replied:
@JoshSmith that's a good idea... but no more great photos of Bill Murray in tests.

WenInCode commented:
I believe I have addressed all the feedback. I rebased as there was a conflict with `develop`. 

This should be ready for review. Thanks to all who have put time into this!

JoshSmith commented:
@WenInCode can you repush to restart the build? Thanks!

JoshSmith commented:
@WenInCode to fix the build, try to change `circle.yml` L30 from `./node_modules/coveralls/bin/coveralls.js < ./coverage/lcov.info` to `COVERAGE=true ./node_modules/coveralls/bin/coveralls.js < ./coverage/lcov.info`

JoshSmith commented:
@WenInCode there are some conflicts now due to some upstream CSS changes merging in #846.

WenInCode commented:
@JoshSmith should be good to review again
